{
 "cells": [
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Check availability data"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 1,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": [
    "#Upload packages\n",
    "import pandas as pd\n",
    "import cobra\n",
    "import alloregfit as arf"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Upload data and metabolic model"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "metadata": {
    "collapsed": false
   },
   "outputs": [
    {
     "data": {
      "text/html": [
       "\n",
       "        <table>\n",
       "            <tr>\n",
       "                <td><strong>Name</strong></td>\n",
       "                <td>iJO1366</td>\n",
       "            </tr><tr>\n",
       "                <td><strong>Memory address</strong></td>\n",
       "                <td>0x01ceb6adfcc0</td>\n",
       "            </tr><tr>\n",
       "                <td><strong>Number of metabolites</strong></td>\n",
       "                <td>1805</td>\n",
       "            </tr><tr>\n",
       "                <td><strong>Number of reactions</strong></td>\n",
       "                <td>2583</td>\n",
       "            </tr><tr>\n",
       "                <td><strong>Objective expression</strong></td>\n",
       "                <td>-1.0*BIOMASS_Ec_iJO1366_core_53p95M_reverse_5c8b1 + 1.0*BIOMASS_Ec_iJO1366_core_53p95M</td>\n",
       "            </tr><tr>\n",
       "                <td><strong>Compartments</strong></td>\n",
       "                <td>c, e, p</td>\n",
       "            </tr>\n",
       "          </table>"
      ],
      "text/plain": [
       "<Model iJO1366 at 0x1ceb6adfcc0>"
      ]
     },
     "execution_count": 2,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "data_dir = \"C:/Users/user/polybox/MASTER/THESIS/3_Karl_data/\"\n",
    "model = cobra.io.load_matlab_model(data_dir+\"iJO1366.mat\")\n",
    "fluxes = pd.read_excel(data_dir+\"fluxes_GS.xlsx\",index_col=\"name\")\n",
    "metabolites = pd.read_excel(data_dir+\"merged_metabolites.xlsx\",index_col=\"name\")\n",
    "proteins = pd.read_excel(data_dir+\"proteome.xlsx\",index_col=\"name\")\n",
    "proteins2 = pd.read_excel(data_dir+\"proteome_bnames.xlsx\",index_col=\"name\")\n",
    "mapping = pd.read_table(data_dir+\"ECOLI_83333_idmapping.dat\",header=None)\n",
    "model"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {
    "collapsed": true
   },
   "source": [
    "## Check data availability for every reaction in model"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Define the function that inputs the data and analyzes how many reactions have enough data:"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 106,
   "metadata": {
    "collapsed": false
   },
   "outputs": [],
   "source": [
    "def enough_data(model,fluxes,proteins,metabolites):\n",
    "    for i in range(len(model.reactions)):\n",
    "        rxn = model.reactions[i]\n",
    "        substrates = rxn.reactants # improvement: consider fluxes\n",
    "        genes = list(rxn.genes) # use gene_reaction_rule to consider and/or\n",
    "\n",
    "        # Check fluxes\n",
    "        flux_bool = False\n",
    "        if any(rxn.id in s for s in list(fluxes.index.values)):\n",
    "            flux_bool = True\n",
    "\n",
    "        # Check proteins\n",
    "        prot_bool = False\n",
    "        for j in range(len(genes)):\n",
    "            prot = mapping[mapping[2]==genes[j].id][0].reset_index()\n",
    "            if prot.empty==0:\n",
    "                gene = list(mapping[(mapping[0]==prot[0][0]) & (mapping[1]=='Gene_Name')][2])\n",
    "                if any(gene[0] in s for s in list(proteins.index.values)):\n",
    "                    prot_bool = True\n",
    "                    break\n",
    "\n",
    "        # Check substrates\n",
    "        met_bool = False\n",
    "        for j in range(len(substrates)):\n",
    "            if any(substrates[j].id[:-2] in s for s in ['h','h2o'])==0:\n",
    "                if any(substrates[j].id[:-2] in s for s in list(metabolites.index.values))==0:\n",
    "                    break\n",
    "        else:\n",
    "            met_bool = True\n",
    "        if flux_bool and met_bool and prot_bool:\n",
    "            print(\"Enough data for %s (%s): %s\" % (rxn.name, rxn.id, rxn.reaction))\n",
    "    "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": [
    "def enough_data_bn(model,fluxes,proteins,metabolites):\n",
    "    for i in range(len(model.reactions)):\n",
    "        rxn = model.reactions[i]\n",
    "        substrates = rxn.reactants # improvement: consider fluxes\n",
    "        genes = list(rxn.genes) # use gene_reaction_rule to consider and/or\n",
    "\n",
    "        # Check fluxes\n",
    "        flux_bool = False\n",
    "        if any(rxn.id in s for s in list(fluxes.index.values)):\n",
    "            flux_bool = True\n",
    "\n",
    "        # Check proteins\n",
    "        prot_bool = False\n",
    "        for j in range(len(genes)):\n",
    "            if any(genes[j].id in s for s in list(proteins.index.values)):\n",
    "                prot_bool = True\n",
    "                break\n",
    "\n",
    "        # Check substrates\n",
    "        met_bool = False\n",
    "        for j in range(len(substrates)):\n",
    "            if any(substrates[j].id[:-2] in s for s in ['h','h2o'])==0:\n",
    "                if any(substrates[j].id[:-2] in s for s in list(metabolites.index.values))==0:\n",
    "                    break\n",
    "        else:\n",
    "            met_bool = True\n",
    "        if flux_bool and met_bool and prot_bool:\n",
    "            print(\"Enough data for %s (%s): %s\" % (rxn.name, rxn.id, rxn.reaction))"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Apply function to our model and Gerosa dataset. Be aware that metabolites, fluxes and proteins need to share same names as in the model in order to be correctly interpreted by the function."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 93,
   "metadata": {
    "collapsed": false
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Enough data for Fructose-bisphosphate aldolase (FBA): fdp_c <=> dhap_c + g3p_c\n",
      "Enough data for Fumarase (FUM): fum_c + h2o_c <=> mal__L_c\n",
      "Enough data for Glucose 6-phosphate dehydrogenase (G6PDH2r): g6p_c + nadp_c <=> 6pgl_c + h_c + nadph_c\n",
      "Enough data for Phosphogluconate dehydrogenase (GND): 6pgc_c + nadp_c --> co2_c + nadph_c + ru5p__D_c\n",
      "Enough data for Malate dehydrogenase (MDH): mal__L_c + nad_c <=> h_c + nadh_c + oaa_c\n",
      "Enough data for NAD transhydrogenase (NADTRHD): nad_c + nadph_c --> nadh_c + nadp_c\n",
      "Enough data for Phosphofructokinase (PFK): atp_c + f6p_c --> adp_c + fdp_c + h_c\n",
      "Enough data for Glucose-6-phosphate isomerase (PGI): g6p_c <=> f6p_c\n",
      "Enough data for Pyruvate kinase (PYK): adp_c + h_c + pep_c --> atp_c + pyr_c\n",
      "Enough data for Transketolase (TKT1): r5p_c + xu5p__D_c <=> g3p_c + s7p_c\n",
      "Enough data for Triose-phosphate isomerase (TPI): dhap_c <=> g3p_c\n"
     ]
    }
   ],
   "source": [
    "enough_data(model,fluxes,proteins,metabolites)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "metadata": {
    "collapsed": false
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Enough data for 23cCMP transport via diffusion (extracellular to periplasm) (23CCMPtex): 23ccmp_e <=> 23ccmp_p\n",
      "Enough data for 23cUMP transport via diffusion (extracellular to periplasm) (23CUMPtex): 23cump_e <=> 23cump_p\n",
      "Enough data for 4-Hydroxy-L-threonine synthase (4HTHRS): h2o_c + phthr_c --> 4hthr_c + pi_c\n",
      "Enough data for Arabinose-5-phosphate isomerase (A5PISO): ru5p__D_c <=> ara5p_c\n",
      "Enough data for 4-aminobutyrate transaminase (ABTA): 4abut_c + akg_c --> glu__L_c + sucsal_c\n",
      "Enough data for 4-aminobutyrate transport via diffusion (extracellular to periplasm) (ABUTtex): 4abut_e <=> 4abut_p\n",
      "Enough data for Acetyl-CoA C-acetyltransferase (ACACT1r): 2.0 accoa_c <=> aacoa_c + coa_c\n",
      "Enough data for Acetyl-CoA C-acyltransferase (butanoyl-CoA) (r) (ACACT2r): accoa_c + btcoa_c <=> 3ohcoa_c + coa_c\n",
      "Enough data for Acetaldehyde dehydrogenase (acetylating) (ACALD): acald_c + coa_c + nad_c <=> accoa_c + h_c + nadh_c\n",
      "Enough data for Acetaldehyde transport via diffusion (extracellular to periplasm) (ACALDtex): acald_e <=> acald_p\n",
      "Enough data for N-Acetyl-D-glucosamine transport via PEP:Pyr PTS  (periplasm) (ACGAptspp): acgam_p + pep_c --> acgam6p_c + pyr_c\n",
      "Enough data for N-Acetyl-D-glucosamine transport via diffusion (extracellular to periplasm) (ACGAtex): acgam_e <=> acgam_p\n",
      "Enough data for Acetylglutamate kinase (ACGK): acglu_c + atp_c --> acg5p_c + adp_c\n",
      "Enough data for N-acetylglutamate synthase (ACGS): accoa_c + glu__L_c --> acglu_c + coa_c + h_c\n",
      "Enough data for 2-aceto-2-hydroxybutanoate synthase (ACHBS): 2obut_c + h_c + pyr_c --> 2ahbut_c + co2_c\n",
      "Enough data for Acetate kinase (ACKr): ac_c + atp_c <=> actp_c + adp_c\n",
      "Enough data for Acetolactate synthase (ACLS): h_c + 2.0 pyr_c --> alac__S_c + co2_c\n",
      "Enough data for N-acetylmuramate transport via PEP:Pyr PTS (periplasm) (ACMUMptspp): acmum_p + pep_c --> acmum6p_c + pyr_c\n",
      "Enough data for N-acetylmuramate transport via diffusion (extracellular to periplasm) (ACMUMtex): acmum_e --> acmum_p\n",
      "Enough data for N-acetylneuraminate transport via diffusion (extracellular to periplasm) (ACNAMtex): acnam_e <=> acnam_p\n",
      "Enough data for N-Acetylneuraminate lyase (ACNML): acnam_c --> acmana_c + pyr_c\n",
      "Enough data for Acyl-CoA dehydrogenase (butanoyl-CoA) (ACOAD1f): btcoa_c + fad_c <=> b2coa_c + fadh2_c\n",
      "Enough data for Aconitase (half-reaction A, Citrate hydro-lyase) (ACONTa): cit_c <=> acon_C_c + h2o_c\n",
      "Enough data for Aconitase (half-reaction B, Isocitrate hydro-lyase) (ACONTb): acon_C_c + h2o_c <=> icit_c\n",
      "Enough data for Acetyl-CoA synthetase (ACS): ac_c + atp_c + coa_c --> accoa_c + amp_c + ppi_c\n",
      "Enough data for Acetate transport via diffusion (extracellular to periplasm) (ACtex): ac_e <=> ac_p\n",
      "Enough data for Adenine transport via diffusion (extracellular to periplasm) (ADEtex): ade_e <=> ade_p\n",
      "Enough data for Adenylate kinase (ADK1): amp_c + atp_c <=> 2.0 adp_c\n",
      "Enough data for Adentylate kinase (GTP) (ADK3): amp_c + gtp_c <=> adp_c + gdp_c\n",
      "Enough data for Adenosine kinase (ADNK1): adn_c + atp_c --> adp_c + amp_c + h_c\n",
      "Enough data for Adenosine hydrolase (ADNUC): adn_c + h2o_c --> ade_c + rib__D_c\n",
      "Enough data for Adenosine transport via diffusion (extracellular to periplasm) (ADNtex): adn_e <=> adn_p\n",
      "Enough data for Adenine phosphoribosyltransferase (ADPT): ade_c + prpp_c --> amp_c + ppi_c\n",
      "Enough data for Adenylsuccinate lyase (ADSL1r): dcamp_c <=> amp_c + fum_c\n",
      "Enough data for Adenylosuccinate synthase (ADSS): asp__L_c + gtp_c + imp_c --> dcamp_c + gdp_c + 2.0 h_c + pi_c\n",
      "Enough data for N-acetylglucosamine-6-phosphate deacetylase (AGDC): acgam6p_c + h2o_c --> ac_c + gam6p_c\n",
      "Enough data for ADP-D-glycero-D-manno-heptose epimerase (AGMHE): adphep_DD_c --> adphep_LD_c\n",
      "Enough data for N-acetyl-g-glutamyl-phosphate reductase (AGPR): acg5sa_c + nadp_c + pi_c <=> acg5p_c + h_c + nadph_c\n",
      "Enough data for Phosphoribosylaminoimidazole carboxylase (mutase rxn) (AIRC3): 5aizc_c <=> 5caiz_c\n",
      "Enough data for 2-Oxogluterate dehydrogenase (AKGDH): akg_c + coa_c + nad_c --> co2_c + nadh_c + succoa_c\n",
      "Enough data for Alpha-ketoglutarate transport via diffusion (extracellular to periplasm) (AKGtex): akg_e <=> akg_p\n",
      "Enough data for L-alanine transaminase (ALATA_L): akg_c + ala__L_c <=> glu__L_c + pyr_c\n",
      "Enough data for Alanine transaminase (ALATA_L2): ala__L_c + pydx5p_c --> pyam5p_c + pyr_c\n",
      "Enough data for L-alanine transport via ABC system (periplasm) (ALAabcpp): ala__L_p + atp_c + h2o_c --> adp_c + ala__L_c + h_c + pi_c\n",
      "Enough data for L-alanine transport via diffusion (extracellular to periplasm) (ALAtex): ala__L_e <=> ala__L_p\n",
      "Enough data for Aldehyde dehydrogenase (acetaldehyde, NADP) (ALDD2y): acald_c + h2o_c + nadp_c --> ac_c + 2.0 h_c + nadph_c\n",
      "Enough data for Aldose reductase (methylglyoxal) (ALR2): h_c + mthgxl_c + nadph_c --> acetol_c + nadp_c\n",
      "Enough data for AMP nucleosidase (AMPN): amp_c + h2o_c --> ade_c + r5p_c\n",
      "Enough data for Alanyl aminopeptidase (cys-gly) (AMPTASECG): cgly_c + h2o_c --> cys__L_c + gly_c\n",
      "Enough data for AMP transport via diffusion (extracellular to periplasm) (AMPtex): amp_e <=> amp_p\n",
      "Enough data for Anthranilate phosphoribosyltransferase (ANPRT): anth_c + prpp_c --> ppi_c + pran_c\n",
      "Enough data for Ap4A synthetase (AP4AS): 2.0 atp_c + h_c --> ap4a_c + ppi_c\n",
      "Enough data for Arginine decarboxylase (ARGDCpp): arg__L_p + h_p --> agm_p + co2_p\n",
      "Enough data for Argininosuccinate synthase (ARGSS): asp__L_c + atp_c + citr__L_c --> amp_c + argsuc_c + h_c + ppi_c\n",
      "Enough data for L-arginine transport via ABC system (periplasm) (ARGabcpp): arg__L_p + atp_c + h2o_c --> adp_c + arg__L_c + h_c + pi_c\n",
      "Enough data for L-arginine transport via diffusion (extracellular to periplasm) (ARGtex): arg__L_e <=> arg__L_p\n",
      "Enough data for Aspartate-semialdehyde dehydrogenase (ASAD): aspsa_c + nadp_c + pi_c <=> 4pasp_c + h_c + nadph_c\n",
      "Enough data for L-ascorbate transport via PEP:Pyr PTS (periplasm) (ASCBptspp): ascb__L_p + pep_c --> ascb6p_c + pyr_c\n",
      "Enough data for L-ascorbate transport via diffusion (extracellular to periplasm) (ASCBtex): ascb__L_e <=> ascb__L_p\n",
      "Enough data for Asparagine synthase (glutamine-hydrolysing) (ASNS1): asp__L_c + atp_c + gln__L_c + h2o_c --> amp_c + asn__L_c + glu__L_c + h_c + ppi_c\n",
      "Enough data for L-asparagine transport via diffusion (extracellular to periplasm) (ASNtex): asn__L_e <=> asn__L_p\n",
      "Enough data for Aspartate 1-decarboxylase (ASP1DC): asp__L_c + h_c --> ala_B_c + co2_c\n",
      "Enough data for Aspartate carbamoyltransferase (ASPCT): asp__L_c + cbp_c --> cbasp_c + h_c + pi_c\n",
      "Enough data for Aspartate kinase (ASPK): asp__L_c + atp_c <=> 4pasp_c + adp_c\n",
      "Enough data for L-aspartate oxidase (ASPO5): asp__L_c + fum_c --> h_c + iasp_c + succ_c\n",
      "Enough data for Aspartate transaminase (ASPTA): akg_c + asp__L_c <=> glu__L_c + oaa_c\n",
      "Enough data for L-aspartate transport via ABC system (periplasm) (ASPabcpp): asp__L_p + atp_c + h2o_c --> adp_c + asp__L_c + h_c + pi_c\n",
      "Enough data for L-aspartate transport via diffusion (extracellular to periplasm) (ASPtex): asp__L_e <=> asp__L_p\n",
      "Enough data for ATP phosphoribosyltransferase (ATPPRT): atp_c + prpp_c --> ppi_c + prbatp_c\n",
      "Enough data for ATP synthase (four protons for one ATP) (periplasm) (ATPS4rpp): adp_c + 4.0 h_p + pi_c <=> atp_c + h2o_c + 3.0 h_c\n",
      "Enough data for Butyrate transport via diffusion (extracellular to periplasm) (BUTtex): but_e <=> but_p\n",
      "Enough data for Cobinamide adenyltransferase (CBIAT): atp_c + cbi_c + h_c <=> adocbi_c + pppi_c\n",
      "Enough data for Cobinamide transport via ton system (extermal) (CBItonex): cbi_e + h_p --> cbi_p + h_c\n",
      "Enough data for L-Cysteinylglycine (Cys-Gly) transport via ABC system (periplasm) (CGLYabcpp): atp_c + cgly_p + h2o_c --> adp_c + cgly_c + h_c + pi_c\n",
      "Enough data for L-Cysteinylglycine transport via diffusion (extracellular to periplasm) (CGLYtex): cgly_e <=> cgly_p\n",
      "Enough data for Citrate transport via diffusion (extracellular to periplasm) (CITtex): cit_e <=> cit_p\n",
      "Enough data for Chloride (Cl-1) transport via diffusion (extracellular to periplasm) (CLtex): cl_e <=> cl_p\n",
      "Enough data for CMP transport via diffusion (extracellular to periplasm) (CMPtex): cmp_e <=> cmp_p\n",
      "Enough data for Chloramphenicol transport via diffusion (extracellular to periplasm) (CMtex): cm_e <=> cm_p\n",
      "Enough data for Chloramphenicol transport via TolC system (CMtpp): cm_p + h_p --> cm_e + h_c\n",
      "Enough data for Cyclic pyranopterin monophosphate synthase (CPMPS): gtp_c + h2o_c --> cpmp_c + ppi_c\n",
      "Enough data for L-carnitine transport via ABC system (periplasm) (CRNabcpp): atp_c + crn_p + h2o_c --> adp_c + crn_c + h_c + pi_c\n",
      "Enough data for L-carnitine transport via diffusion (extracellular to periplasm) (CRNtex): crn_e <=> crn_p\n",
      "Enough data for Cytosine deaminase (CSND): csn_c + h2o_c + h_c --> nh4_c + ura_c\n",
      "Enough data for Cytosine transport via diffusion (extracellular to periplasm) (CSNtex): csn_e <=> csn_p\n",
      "Enough data for CTP synthase (glutamine) (CTPS2): atp_c + gln__L_c + h2o_c + utp_c --> adp_c + ctp_c + glu__L_c + 2.0 h_c + pi_c\n",
      "Enough data for Copper (Cu +1) ABC transporter (periplasm) (CU1abcpp): atp_c + cu_c + h2o_c --> adp_c + cu_p + h_c + pi_c\n",
      "Enough data for Copper transport out via proton antiport (CUt3): cu_c + h_e --> cu_e + h_c\n",
      "Enough data for Copper (Cu+1) transport via diffusion (extracellular to periplasm) (CUtex): cu_e <=> cu_p\n",
      "Enough data for Cyanate transport via diffusion (extracellular to periplasm) (CYNTtex): cynt_e <=> cynt_p\n",
      "Enough data for Cysteine Desulfhydrase (CYSDS): cys__L_c + h2o_c --> h2s_c + nh4_c + pyr_c\n",
      "Enough data for L-cysteine transport via diffusion (extracellular to periplasm) (CYStex): cys__L_e <=> cys__L_p\n",
      "Enough data for Cytidine hydrolase (CYTDH): cytd_c + h2o_c --> csn_c + rib__D_c\n",
      "Enough data for Cytidine transport via diffusion (extracellular to periplasm) (CYTDtex): cytd_e <=> cytd_p\n",
      "Enough data for Cytidylate kinase (CMP) (CYTK1): atp_c + cmp_c <=> adp_c + cdp_c\n",
      "Enough data for Cytidylate kinase (dCMP) (CYTK2): atp_c + dcmp_c <=> adp_c + dcdp_c\n",
      "Enough data for Deoxyadenylate kinase (DADK): atp_c + damp_c <=> adp_c + dadp_c\n",
      "Enough data for Deoxyadenosine transport via diffusion (extracellular to periplasm) (DADNtex): dad_2_e <=> dad_2_p\n",
      "Enough data for DAMP transport via diffusion (extracellular to periplasm) (DAMPtex): damp_e <=> damp_p\n",
      "Enough data for Decanoate transport via diffusion (extracellular to periplasm) (DCAtex): dca_e <=> dca_p\n",
      "Enough data for DCMP transport via diffusion (extracellular to periplasm) (DCMPtex): dcmp_e <=> dcmp_p\n",
      "Enough data for Deoxycytidine transport via diffusion (extracellular to periplasm) (DCYTtex): dcyt_e <=> dcyt_p\n",
      "Enough data for 3-deoxy-D-arabino-heptulosonate 7-phosphate synthetase (DDPA): e4p_c + h2o_c + pep_c --> 2dda7p_c + pi_c\n",
      "Enough data for Dihydroxyacetone phosphotransferase (DHAPT): dha_c + pep_c --> dhap_c + pyr_c\n",
      "Enough data for Dihydroxyacetone transport via diffusion (extracellular to periplasm) (DHAtex): dha_e <=> dha_p\n",
      "Enough data for 2,3-dihydroxybenzoate adenylate synthase (DHBS): 23dhb_c + atp_c + h_c --> 23dhba_c + ppi_c\n",
      "Enough data for Dihydrodipicolinate synthase (DHDPS): aspsa_c + pyr_c --> 23dhdp_c + 2.0 h2o_c + h_c\n",
      "Enough data for Dihydroorotase (DHORTS): dhor__S_c + h2o_c <=> cbasp_c + h_c\n",
      "Enough data for 3-dehydroquinate synthase (DHQS): 2dda7p_c --> 3dhq_c + pi_c\n",
      "Enough data for 3-dehydroquinate dehydratase, irreversible (DHQTi): 3dhq_c --> 3dhsk_c + h2o_c\n",
      "Enough data for DIMP transport via diffusion (extracellular to periplasm) (DIMPtex): dimp_e <=> dimp_p\n",
      "Enough data for Dopamine transport via diffusion (extracellular to periplasm) (DOPAtex): dopa_e <=> dopa_p\n",
      "Enough data for DUTP diphosphatase (DUTPDP): dutp_c + h2o_c --> dump_c + h_c + ppi_c\n",
      "Enough data for 1-deoxy-D-xylulose 5-phosphate synthase (DXPS): g3p_c + h_c + pyr_c --> co2_c + dxyl5p_c\n",
      "Enough data for Erythrose 4-phosphate dehydrogenase (E4PD): e4p_c + h2o_c + nad_c <=> 4per_c + 2.0 h_c + nadh_c\n",
      "Enough data for Fructose 6-phosphate transport via diffusion (extracellular to periplasm) (F6Ptex): f6p_e <=> f6p_p\n",
      "Enough data for Fatty-acid-CoA ligase (decanoate transport via vectoral Co-A coupling) (FACOAL100t2pp): atp_c + coa_c + dca_p + h_p --> amp_c + dcacoa_c + h_c + ppi_c\n",
      "Enough data for Fatty-acid-CoA ligase (hexadecanoate transport via vectoral Co-A coupling) (FACOAL160t2pp): atp_c + coa_c + h_p + hdca_p --> amp_c + h_c + pmtcoa_c + ppi_c\n",
      "Enough data for Fatty-acid-CoA ligase (hexanoate transport via vectoral Co-A coupling) (FACOAL60t2pp): atp_c + coa_c + h_p + hxa_p --> amp_c + h_c + hxcoa_c + ppi_c\n",
      "Enough data for FAD reductase (FADRx): fad_c + h_c + nadh_c --> fadh2_c + nad_c\n",
      "Enough data for FAD reductase (FADRx2): fad_c + h_c + nadph_c --> fadh2_c + nadp_c\n",
      "Enough data for Formaldehyde dehydrogenase (FALDH2): hmgth_c + nad_c <=> Sfglutth_c + h_c + nadh_c\n",
      "Enough data for Fructose-bisphosphate aldolase (FBA): fdp_c <=> dhap_c + g3p_c\n",
      "Enough data for Fructose-bisphosphatase (FBP): fdp_c + h2o_c --> f6p_c + pi_c\n",
      "Enough data for Enterochelin transport (secretion periplasm) (FEENTERtex): enter_p + h_p --> enter_e + h_c\n",
      "Enough data for Flavin reductase (FLVR): h_c + nadph_c + ribflv_c --> nadp_c + rbflvrd_c\n",
      "Enough data for Flavin reductase (NAD) (FLVRx): h_c + nadh_c + ribflv_c --> nad_c + rbflvrd_c\n",
      "Enough data for Fumarase (FUM): fum_c + h2o_c <=> mal__L_c\n",
      "Enough data for Fumarate transport via diffusion (extracellular to periplasm) (FUMtex): fum_e <=> fum_p\n",
      "Enough data for Glycero-3-phosphocholine transport via diffusion (extracellular to periplasm) (G3PCtex): g3pc_e <=> g3pc_p\n",
      "Enough data for Glycerol-3-phosphate dehydrogenase (NADP) (G3PD2): glyc3p_c + nadp_c <=> dhap_c + h_c + nadph_c\n",
      "Enough data for Glycero-3-phosphoethanolamine transport via diffusion (extracellular to periplasm) (G3PEtex): g3pe_e <=> g3pe_p\n",
      "Enough data for Glucose 6-phosphate dehydrogenase (G6PDH2r): g6p_c + nadp_c <=> 6pgl_c + h_c + nadph_c\n",
      "Enough data for Glucose 6-phosphate transport via diffusion (extracellular to periplasm) (G6Ptex): g6p_e <=> g6p_p\n",
      "Enough data for Galactokinase (GALKr): atp_c + gal_c <=> adp_c + gal1p_c + h_c\n",
      "Enough data for D-galacturonate transport via diffusion (extracellular to periplasm) (GALURtex): galur_e <=> galur_p\n",
      "Enough data for D-galactose transport via ABC system (periplasm) (GALabcpp): atp_c + gal_p + h2o_c --> adp_c + gal_c + h_c + pi_c\n",
      "Enough data for D-galactose transport via diffusion (extracellular to periplasm) (GALtex): gal_e <=> gal_p\n",
      "Enough data for D-glucosamine 6-phosphate transport via diffusion (extracellular to periplasm) (GAMAN6Ptex): gam6p_e <=> gam6p_p\n",
      "Enough data for D-glucosamine transport via PEP:Pyr PTS (periplasm) (GAMptspp): gam_p + pep_c --> gam6p_c + pyr_c\n",
      "Enough data for D-glucosamine transport via diffusion (extracellular to periplasm) (GAMtex): gam_e <=> gam_p\n",
      "Enough data for Glyceraldehyde-3-phosphate dehydrogenase (GAPD): g3p_c + nad_c + pi_c <=> 13dpg_c + h_c + nadh_c\n",
      "Enough data for GDP diphosphokinase (GDPDPK): atp_c + gdp_c --> amp_c + h_c + ppgpp_c\n",
      "Enough data for GDP transport via diffusion (extracellular to periplasm) (GDPtex): gdp_e <=> gdp_p\n",
      "Enough data for Glutamine-fructose-6-phosphate transaminase (GF6PTA): f6p_c + gln__L_c --> gam6p_c + glu__L_c\n",
      "Enough data for Glycine hydroxymethyltransferase, reversible (GHMT2r): ser__L_c + thf_c <=> gly_c + h2o_c + mlthf_c\n",
      "Enough data for Guanylate kinase (GMP:ATP) (GK1): atp_c + gmp_c <=> adp_c + gdp_c\n",
      "Enough data for D-gluconate transport via diffusion (extracellular to periplasm) (GLCNtex): glcn_e <=> glcn_p\n",
      "Enough data for D-glucose transport via ABC system (periplasm) (GLCabcpp): atp_c + glc__D_p + h2o_c --> adp_c + glc__D_c + h_c + pi_c\n",
      "Enough data for D-glucose transport via PEP:Pyr PTS (periplasm) (GLCptspp): glc__D_p + pep_c --> g6p_c + pyr_c\n",
      "Enough data for L-glutamine transport via ABC system (periplasm) (GLNabcpp): atp_c + gln__L_p + h2o_c --> adp_c + gln__L_c + h_c + pi_c\n",
      "Enough data for L-glutamine transport via diffusion (extracellular to periplasm) (GLNtex): gln__L_e <=> gln__L_p\n",
      "Enough data for Glutamate 5-kinase (GLU5K): atp_c + glu__L_c --> adp_c + glu5p_c\n",
      "Enough data for Gamma-glutamylcysteine synthetase (GLUCYS): atp_c + cys__L_c + glu__L_c --> adp_c + glucys_c + h_c + pi_c\n",
      "Enough data for Glutamate dehydrogenase (NADP) (GLUDy): glu__L_c + h2o_c + nadp_c <=> akg_c + h_c + nadph_c + nh4_c\n",
      "Enough data for Glutamine phosphoribosyldiphosphate amidotransferase (GLUPRT): gln__L_c + h2o_c + prpp_c --> glu__L_c + ppi_c + pram_c\n",
      "Enough data for Glutamate synthase (NADPH) (GLUSy): akg_c + gln__L_c + h_c + nadph_c --> 2.0 glu__L_c + nadp_c\n",
      "Enough data for L-glutamate transport via ABC system (periplasm) (GLUabcpp): atp_c + glu__L_p + h2o_c --> adp_c + glu__L_c + h_c + pi_c\n",
      "Enough data for L-glutamate transport via diffusion (extracellular to periplasm) (GLUtex): glu__L_e <=> glu__L_p\n",
      "Enough data for Glycerol-3-phosphate transport via diffusion (extracellular to periplasm) (GLYC3Ptex): glyc3p_e <=> glyc3p_p\n",
      "Enough data for D-glycerate transport via diffusion (extracellular to periplasm) (GLYCAtex): glyc__R_e <=> glyc__R_p\n",
      "Enough data for Glycine Cleavage System (GLYCL): gly_c + nad_c + thf_c --> co2_c + mlthf_c + nadh_c + nh4_c\n",
      "Enough data for Glycolate dehydrogenase (NAD) (GLYCLTDx): glx_c + h_c + nadh_c --> glyclt_c + nad_c\n",
      "Enough data for Glycolate dehydrogenase (NADP) (GLYCLTDy): glx_c + h_c + nadph_c --> glyclt_c + nadp_c\n",
      "Enough data for Glycolate transport via diffusion (extracellular to periplasm) (GLYCLTtex): glyclt_e <=> glyclt_p\n",
      "Enough data for Glycerol transport via diffusion (extracellular to periplasm) (GLYCtex): glyc_e <=> glyc_p\n",
      "Enough data for Glycerol transport via channel (periplasm) (GLYCtpp): glyc_c <=> glyc_p\n",
      "Enough data for Glycerol kinase (GLYK): atp_c + glyc_c --> adp_c + glyc3p_c + h_c\n",
      "Enough data for Glycine transport via diffusion (extracellular to periplasm) (GLYtex): gly_e <=> gly_p\n",
      "Enough data for GMP reductase (GMPR): gmp_c + 2.0 h_c + nadph_c --> imp_c + nadp_c + nh4_c\n",
      "Enough data for GMP transport via diffusion (extracellular to periplasm) (GMPtex): gmp_e <=> gmp_p\n",
      "Enough data for Phosphogluconate dehydrogenase (GND): 6pgc_c + nadp_c --> co2_c + nadph_c + ru5p__D_c\n",
      "Enough data for Glutathionylspermidine amidase (GSPMDA): gtspmd_c + h2o_c --> gthrd_c + spmd_c\n",
      "Enough data for Glutathionylspermidine synthetase (GSPMDS): atp_c + gthrd_c + spmd_c --> adp_c + gtspmd_c + h_c + pi_c\n",
      "Enough data for Glutathione (ox) transport via diffusion (extracellular to periplasm) (GTHOXtex): gthox_e <=> gthox_p\n",
      "Enough data for Glutathione oxidoreductase (GTHOr): gthox_c + h_c + nadph_c <=> 2.0 gthrd_c + nadp_c\n",
      "Enough data for Reduced glutathione via ABC system (periplasm) (GTHRDabcpp): atp_c + gthrd_p + h2o_c --> adp_c + gthrd_c + h_c + pi_c\n",
      "Enough data for Glutathione transport via diffusion (extracellular to periplasm) (GTHRDtex): gthrd_e <=> gthrd_p\n",
      "Enough data for GTP cyclohydrolase I (GTPCI): gtp_c + h2o_c --> ahdt_c + for_c + h_c\n",
      "Enough data for GTP diphosphokinase (GTPDPK): atp_c + gtp_c --> amp_c + gdptp_c + h_c\n",
      "Enough data for GTP transport via diffusion (extracellular to periplasm) (GTPtex): gtp_e <=> gtp_p\n",
      "Enough data for Guanine phosphoribosyltransferase (GUAPRT): gua_c + prpp_c --> gmp_c + ppi_c\n",
      "Enough data for Guanine transport via diffusion (extracellular to periplasm) (GUAtex): gua_e <=> gua_p\n",
      "Enough data for H2O transport via diffusion (extracellular to periplasm) (H2Otex): h2o_e <=> h2o_p\n",
      "Enough data for Hydrogen transport via diffusion (extracellular to periplasm) (H2tex): h2_e <=> h2_p\n",
      "Enough data for Hexadecanoate transport via facilitated irreversible diffusion (extracellular to periplasm) (HDCAtexi): hdca_e --> hdca_p\n",
      "Enough data for Hexokinase (D-glucose:ATP) (HEX1): atp_c + glc__D_c --> adp_c + g6p_c + h_c\n",
      "Enough data for Histidinol-phosphatase (HISTP): h2o_c + hisp_c --> histd_c + pi_c\n",
      "Enough data for L-histidine transport via ABC system (periplasm) (HISabcpp): atp_c + h2o_c + his__L_p --> adp_c + h_c + his__L_c + pi_c\n",
      "Enough data for L-histidine transport via diffusion (extracellular to periplasm) (HIStex): his__L_e <=> his__L_p\n",
      "Enough data for Hydroxypyruvate reductase (NADH) (HPYRRx): h_c + hpyr_c + nadh_c --> glyc__R_c + nad_c\n",
      "Enough data for Hydroxypyruvate reductase (NADPH) (HPYRRy): h_c + hpyr_c + nadph_c --> glyc__R_c + nadp_c\n",
      "Enough data for Hexanoate transport via diffusion (extracellular to periplasm) (HXAtex): hxa_e <=> hxa_p\n",
      "Enough data for Hypoxanthine phosphoribosyltransferase (Hypoxanthine) (HXPRT): hxan_c + prpp_c --> imp_c + ppi_c\n",
      "Enough data for Hypoxanthine transport via diffusion (extracellular to periplasm) (HYXNtex): hxan_e <=> hxan_p\n",
      "Enough data for Proton transport via diffusion (extracellular to periplasm) (Htex): h_e <=> h_p\n",
      "Enough data for IMP cyclohydrolase (IMPC): h2o_c + imp_c <=> fprica_c\n",
      "Enough data for IMP dehydrogenase (IMPD): h2o_c + imp_c + nad_c --> h_c + nadh_c + xmp_c\n",
      "Enough data for IMP transport via diffusion (extracellular to periplasm) (IMPtex): imp_e <=> imp_p\n",
      "Enough data for Indole transport via proton symport, irreversible (periplasm) (INDOLEt2pp): h_c + indole_c --> h_p + indole_p\n",
      "Enough data for Indole transport via diffusion (extracellular to periplasm) (INDOLEtex): indole_e <=> indole_p\n",
      "Enough data for Inosine hydrolase (INSH): h2o_c + ins_c --> hxan_c + rib__D_c\n",
      "Enough data for Inosine transport via diffusion (extracellular to periplasm) (INStex): ins_e <=> ins_p\n",
      "Enough data for 2-isopropylmalate hydratase (IPPMIb): 2ippm_c + h2o_c <=> 3c3hmp_c\n",
      "Enough data for 2-isopropylmalate synthase (IPPS): 3mob_c + accoa_c + h2o_c --> 3c3hmp_c + coa_c + h_c\n",
      "Enough data for 3-deoxy-manno-octulosonate cytidylyltransferase (KDOCT2): ctp_c + kdo_c --> ckdo_c + ppi_c\n",
      "Enough data for 3-deoxy-manno-octulosonate-8-phosphatase (KDOPP): h2o_c + kdo8p_c --> kdo_c + pi_c\n",
      "Enough data for Potassium ABC transporter (periplasm) (Kabcpp): atp_c + h2o_c + k_p --> adp_c + h_c + k_c + pi_c\n",
      "Enough data for Potassium transport via diffusion (extracellular to periplasm) (Ktex): k_e <=> k_p\n",
      "Enough data for Leucine transaminase (irreversible) (LEUTAi): 4mop_c + glu__L_c --> akg_c + leu__L_c\n",
      "Enough data for L-leucine transport via ABC system (periplasm) (LEUabcpp): atp_c + h2o_c + leu__L_p --> adp_c + h_c + leu__L_c + pi_c\n",
      "Enough data for L-leucine transport via diffusion (extracellular to periplasm) (LEUtex): leu__L_e <=> leu__L_p\n",
      "Enough data for L-serine dehydrogenase (LSERDHr): nadp_c + ser__L_c <=> 2amsa_c + h_c + nadph_c\n",
      "Enough data for L-lysine transport via ABC system (periplasm) (LYSabcpp): atp_c + h2o_c + lys__L_p --> adp_c + h_c + lys__L_c + pi_c\n",
      "Enough data for L-lysine transport via diffusion (extracellular to periplasm) (LYStex): lys__L_e <=> lys__L_p\n",
      "Enough data for Mannitol-1-phosphate 5-dehydrogenase (M1PD): mnl1p_c + nad_c <=> f6p_c + h_c + nadh_c\n",
      "Enough data for Malate synthase (MALS): accoa_c + glx_c + h2o_c --> coa_c + h_c + mal__L_c\n",
      "Enough data for Malate transport via diffusion (extracellular to periplasm) (MALtex): mal__L_e <=> mal__L_p\n",
      "Enough data for D-mannose transport via PEP:Pyr PTS (periplasm) (MANptspp): man_p + pep_c --> man6p_c + pyr_c\n",
      "Enough data for D-mannose transport via diffusion (extracellular to periplasm) (MANtex): man_e <=> man_p\n",
      "Enough data for Malate dehydrogenase (MDH): mal__L_c + nad_c <=> h_c + nadh_c + oaa_c\n",
      "Enough data for Malic enzyme (NAD) (ME1): mal__L_c + nad_c --> co2_c + nadh_c + pyr_c\n",
      "Enough data for Malic enzyme (NADP) (ME2): mal__L_c + nadp_c --> co2_c + nadph_c + pyr_c\n",
      "Enough data for Methionine adenosyltransferase (METAT): atp_c + h2o_c + met__L_c --> amet_c + pi_c + ppi_c\n",
      "Enough data for L-methionine transport via ABC system (periplasm) (METabcpp): atp_c + h2o_c + met__L_p --> adp_c + h_c + met__L_c + pi_c\n",
      "Enough data for L-methionine transport via diffusion (extracellular to periplasm) (METtex): met__L_e <=> met__L_p\n",
      "Enough data for Methylglyoxal synthase (MGSA): dhap_c --> mthgxl_c + pi_c\n",
      "Enough data for Mannitol transport via PEP:Pyr PTS (periplasm) (MNLptspp): mnl_p + pep_c --> mnl1p_c + pyr_c\n",
      "Enough data for Mannitol transport via diffusion (extracellular to periplasm) (MNLtex): mnl_e <=> mnl_p\n",
      "Enough data for N-acetylornithine deacetylase (NACODA): acg5sa_c + h2o_c --> ac_c + glu5sa_c\n",
      "Enough data for Nicotinic acid transport via diffusion (extracellular to periplasm) (NACtex): nac_e <=> nac_p\n",
      "Enough data for NAD diphosphatase (NADDP): h2o_c + nad_c --> amp_c + 2.0 h_c + nmn_c\n",
      "Enough data for NAD transhydrogenase (NADTRHD): nad_c + nadph_c --> nadh_c + nadp_c\n",
      "Enough data for Nicotinic acid mononucleotide pyrophosphorylase (NAMNPP): atp_c + h2o_c + nac_c + prpp_c --> adp_c + nicrnt_c + pi_c + ppi_c\n",
      "Enough data for Nucleoside-diphosphate kinase (ATP:GDP) (NDPK1): atp_c + gdp_c <=> adp_c + gtp_c\n",
      "Enough data for Nucleoside-diphosphate kinase (ATP:UDP) (NDPK2): atp_c + udp_c <=> adp_c + utp_c\n",
      "Enough data for Nucleoside-diphosphate kinase (ATP:CDP) (NDPK3): atp_c + cdp_c <=> adp_c + ctp_c\n",
      "Enough data for Nucleoside-diphosphate kinase (ATP:dTDP) (NDPK4): atp_c + dtdp_c <=> adp_c + dttp_c\n",
      "Enough data for Nucleoside-diphosphate kinase (ATP:dCDP) (NDPK7): atp_c + dcdp_c <=> adp_c + dctp_c\n",
      "Enough data for Nucleoside-diphosphate kinase (ATP:dADP) (NDPK8): atp_c + dadp_c <=> adp_c + datp_c\n",
      "Enough data for Nitrate transport via diffusion (extracellular to periplasm) (NO3tex): no3_e <=> no3_p\n",
      "Enough data for Nitric oxide transport via diffusion (extracellular to periplasm) (NOtex): no_e <=> no_p\n",
      "Enough data for 5'-nucleotidase (IMP) (NTD11pp): h2o_p + imp_p --> ins_p + pi_p\n",
      "Enough data for 5'-nucleotidase (dIMP) (periplasm) (NTD12pp): dimp_p + h2o_p --> din_p + pi_p\n",
      "Enough data for 5'-nucleotidase (UMP) (NTD2pp): h2o_p + ump_p --> pi_p + uri_p\n",
      "Enough data for 5'-nucleotidase (dCMP) (NTD3pp): dcmp_p + h2o_p --> dcyt_p + pi_p\n",
      "Enough data for 5'-nucleotidase (CMP) (NTD4pp): cmp_p + h2o_p --> cytd_p + pi_p\n",
      "Enough data for 5'-nucleotidase (dAMP) (NTD6pp): damp_p + h2o_p --> dad_2_p + pi_p\n",
      "Enough data for 5'-nucleotidase (AMP) (NTD7pp): amp_p + h2o_p --> adn_p + pi_p\n",
      "Enough data for 5'-nucleotidase (GMP) (NTD9pp): gmp_p + h2o_p --> gsn_p + pi_p\n",
      "Enough data for Nucleoside-triphosphatase (ATP) (NTP1): atp_c + h2o_c --> adp_c + h_c + pi_c\n",
      "Enough data for Nucleoside-triphosphatase (GTP) (NTP3): gtp_c + h2o_c --> gdp_c + h_c + pi_c\n",
      "Enough data for Nucleoside-triphosphatase (CTP) (NTP5): ctp_c + h2o_c --> cdp_c + h_c + pi_c\n",
      "Enough data for 2-Oxobutanoate formate lyase (OBTFL): 2obut_c + coa_c --> for_c + ppcoa_c\n",
      "Enough data for Ornithine carbamoyltransferase (OCBT): cbp_c + orn_c <=> citr__L_c + h_c + pi_c\n",
      "Enough data for Ornithine transport via ABC system (periplasm) (ORNabcpp): atp_c + h2o_c + orn_p --> adp_c + h_c + orn_c + pi_c\n",
      "Enough data for Ornithine transport via diffusion (extracellular to periplasm) (ORNtex): orn_e <=> orn_p\n",
      "Enough data for Phenethylacetaldehyde transport via diffusion (extracellular to periplasm) (PACALDtex): pacald_e <=> pacald_p\n",
      "Enough data for Pyruvate dehydrogenase (PDH): coa_c + nad_c + pyr_c --> accoa_c + co2_c + nadh_c\n",
      "Enough data for Phosphofructokinase (PFK): atp_c + f6p_c --> adp_c + fdp_c + h_c\n",
      "Enough data for Phosphofructokinase (s7p) (PFK_3): atp_c + s7p_c --> adp_c + h_c + s17bp_c\n",
      "Enough data for Pyruvate formate lyase (PFL): coa_c + pyr_c --> accoa_c + for_c\n",
      "Enough data for Phosphoglycerate dehydrogenase (PGCD): 3pg_c + nad_c --> 3php_c + h_c + nadh_c\n",
      "Enough data for Glucose-6-phosphate isomerase (PGI): g6p_c <=> f6p_c\n",
      "Enough data for Phosphoglycerate kinase (PGK): 3pg_c + atp_c <=> 13dpg_c + adp_c\n",
      "Enough data for 6-phosphogluconolactonase (PGL): 6pgl_c + h2o_c --> 6pgc_c + h_c\n",
      "Enough data for Phenylalanine transaminase (PHETA1): akg_c + phe__L_c <=> glu__L_c + phpyr_c\n",
      "Enough data for L-phenylalanine transport via diffusion (extracellular to periplasm) (PHEtex): phe__L_e <=> phe__L_p\n",
      "Enough data for Phosphate transport via diffusion (extracellular to periplasm) (PItex): pi_e <=> pi_p\n",
      "Enough data for Phosphomethylpyrimidine kinase (PMPK): 4ampm_c + atp_c --> 2mahmp_c + adp_c\n",
      "Enough data for Pantothenate transport via diffusion (extracellular to periplasm) (PNTOtex): pnto__R_e <=> pnto__R_p\n",
      "Enough data for Inorganic diphosphatase (PPA): h2o_c + ppi_c --> h_c + 2.0 pi_c\n",
      "Enough data for Porphobilinogen synthase (PPBNGS): 2.0 5aop_c --> 2.0 h2o_c + h_c + ppbng_c\n",
      "Enough data for Polyphosphate kinase (PPK2): atp_c + ppi_c <=> adp_c + pppi_c\n",
      "Enough data for Polyphosphate kinase (PPK): atp_c + pi_c <=> adp_c + ppi_c\n",
      "Enough data for Phosphopentomutase (PPM): r1p_c <=> r5p_c\n",
      "Enough data for Phosphopentomutase 2 (deoxyribose) (PPM2): 2dr1p_c <=> 2dr5p_c\n",
      "Enough data for Phosphoenolpyruvate synthase (PPS): atp_c + h2o_c + pyr_c --> amp_c + 2.0 h_c + pep_c + pi_c\n",
      "Enough data for Phosphoribosylglycinamide synthase (PRAGSr): atp_c + gly_c + pram_c <=> adp_c + gar_c + h_c + pi_c\n",
      "Enough data for Phosphoribosylaminoimidazolesuccinocarboxamide synthase (PRASCSi): 5aizc_c + asp__L_c + atp_c --> 25aics_c + adp_c + h_c + pi_c\n",
      "Enough data for Phosphoribosylformylglycinamidine synthase (PRFGS): atp_c + fgam_c + gln__L_c + h2o_c --> adp_c + fpram_c + glu__L_c + h_c + pi_c\n",
      "Enough data for L-proline transport via ABC system (periplasm) (PROabcpp): atp_c + h2o_c + pro__L_p --> adp_c + h_c + pi_c + pro__L_c\n",
      "Enough data for L-proline transport via diffusion (extracellular to periplasm) (PROtex): pro__L_e <=> pro__L_p\n",
      "Enough data for Phosphoribosylpyrophosphate synthetase (PRPPS): atp_c + r5p_c <=> amp_c + h_c + prpp_c\n",
      "Enough data for 3-phosphoshikimate 1-carboxyvinyltransferase (PSCVT): pep_c + skm5p_c <=> 3psme_c + pi_c\n",
      "Enough data for Phosphoserine transaminase (PSERT): 3php_c + glu__L_c --> akg_c + pser__L_c\n",
      "Enough data for Phospho-L-serine transport via diffusion (extracellular to periplasm) (PSERtex): pser__L_e <=> pser__L_p\n",
      "Enough data for Phosphoserine phosphatase (L-serine) (PSP_L): h2o_c + pser__L_c --> pi_c + ser__L_c\n",
      "Enough data for Phospho-L-serine phosphatase (periplasmic) (PSP_Lpp): h2o_p + pser__L_p --> pi_p + ser__L_p\n",
      "Enough data for Phosphate acetyltransferase (PTA2): pi_c + ppcoa_c --> coa_c + ppap_c\n",
      "Enough data for Phosphotransacetylase (PTAr): accoa_c + pi_c <=> actp_c + coa_c\n",
      "Enough data for Purine-nucleoside phosphorylase (Adenosine) (PUNP1): adn_c + pi_c <=> ade_c + r1p_c\n",
      "Enough data for Purine-nucleoside phosphorylase (Deoxyadenosine) (PUNP2): dad_2_c + pi_c <=> 2dr1p_c + ade_c\n",
      "Enough data for Purine-nucleoside phosphorylase (Inosine) (PUNP5): ins_c + pi_c <=> hxan_c + r1p_c\n",
      "Enough data for Pyridoxal kinase (PYDXK): atp_c + pydx_c --> adp_c + h_c + pydx5p_c\n",
      "Enough data for Pyridoxine kinase (PYDXNK): atp_c + pydxn_c --> adp_c + h_c + pdx5p_c\n",
      "Enough data for Pyridoxine transport via diffusion (extracellular) (PYDXNtex): pydxn_e <=> pydxn_p\n",
      "Enough data for Pyridoxal transport via diffusion (extracellular) (PYDXtex): pydx_e <=> pydx_p\n",
      "Enough data for Pyruvate kinase (PYK): adp_c + h_c + pep_c --> atp_c + pyr_c\n",
      "Enough data for Pyrimidine-nucleoside phosphorylase (uracil) (PYNP2r): pi_c + uri_c <=> r1p_c + ura_c\n",
      "Enough data for Pyruvate transport via diffusion (extracellular to periplasm) (PYRtex): pyr_e <=> pyr_p\n",
      "Enough data for Quinate transport via diffusion (extracellular to periplasm) (QUIN2tex): quin_e <=> quin_p\n",
      "Enough data for Ribose 5-phosphate phosphatase (R5PPpp): h2o_p + r5p_p --> pi_p + rib__D_p\n",
      "Enough data for Ribose 5-phosphate transport via diffusion (extracellular to periplasm) (R5Ptex): r5p_e <=> r5p_p\n",
      "Enough data for Ribokinase (RBK): atp_c + rib__D_c --> adp_c + h_c + r5p_c\n",
      "Enough data for S-ribosylhomocysteine cleavage enzyme (RHCCE): rhcys_c --> dhptd_c + hcys__L_c\n",
      "Enough data for D-ribose transport via ABC system (periplasm) (RIBabcpp): atp_c + h2o_c + rib__D_p --> adp_c + h_c + pi_c + rib__D_c\n",
      "Enough data for Ribose transport via diffusion (extracellular to periplasm) (RIBtex): rib__D_e <=> rib__D_p\n",
      "Enough data for L-rhamnose transport via diffusion (extracellular to periplasm) (RMNtex): rmn_e <=> rmn_p\n",
      "Enough data for Ribulose 5-phosphate 3-epimerase (RPE): ru5p__D_c <=> xu5p__D_c\n",
      "Enough data for Ribose-5-phosphate isomerase (RPI): r5p_c <=> ru5p__D_c\n",
      "Enough data for Sedoheptulose 7-phosphate isomerase (S7PI): s7p_c --> gmhep7p_c\n",
      "Enough data for Sulfate adenyltransferase (SADT2): atp_c + gtp_c + h2o_c + so4_c --> aps_c + gdp_c + pi_c + ppi_c\n",
      "Enough data for Serine O-acetyltransferase (SERAT): accoa_c + ser__L_c <=> acser_c + coa_c\n",
      "Enough data for L-serine transport via diffusion (extracellular to periplasm) (SERtex): ser__L_e <=> ser__L_p\n",
      "Enough data for Shikimate dehydrogenase (SHK3Dr): 3dhsk_c + h_c + nadph_c <=> nadp_c + skm_c\n",
      "Enough data for Shikimate kinase (SHKK): atp_c + skm_c --> adp_c + h_c + skm5p_c\n",
      "Enough data for O-succinylhomoserine lyase (L-cysteine) (SHSL1): cys__L_c + suchms_c --> cyst__L_c + h_c + succ_c\n",
      "Enough data for Shikimate transport via diffusion (extracellular to periplasm) (SKMtex): skm_e <=> skm_p\n",
      "Enough data for Sulfite transport via diffusion (extracellular to periplasm) (SO3tex): so3_e <=> so3_p\n",
      "Enough data for Sulfate transport via diffusion (extracellular to periplasm) (SO4tex): so4_e <=> so4_p\n",
      "Enough data for Spermidine transport via ABC system (periplasm) (SPMDabcpp): atp_c + h2o_c + spmd_p --> adp_c + h_c + pi_c + spmd_c\n",
      "Enough data for Spermidine transport via diffusion (extracellular to periplasm) (SPMDtex): spmd_e <=> spmd_p\n",
      "Enough data for Succinate transport via diffusion (extracellular to periplasm) (SUCCtex): succ_e <=> succ_p\n",
      "Enough data for Succinyl-CoA synthetase (ADP-forming) (SUCOAS): atp_c + coa_c + succ_c <=> adp_c + pi_c + succoa_c\n",
      "Enough data for Sulfite reductase (NADPH2) (SULR): 5.0 h_c + 3.0 nadph_c + so3_c --> 3.0 h2o_c + h2s_c + 3.0 nadp_c\n",
      "Enough data for Sulfate transport via ABC system (periplasm) (SULabcpp): atp_c + h2o_c + so4_p --> adp_c + h_c + pi_c + so4_c\n",
      "Enough data for Transaldolase (TALA): g3p_c + s7p_c <=> e4p_c + f6p_c\n",
      "Enough data for Thiocyanate transport via diffusion (extracellular to periplasm) (TCYNTtex): tcynt_e <=> tcynt_p\n",
      "Enough data for NAD(P) transhydrogenase (periplasm) (THD2pp): 2.0 h_p + nadh_c + nadp_c --> 2.0 h_c + nad_c + nadph_c\n",
      "Enough data for Tetrahydrodipicolinate succinylase (THDPS): h2o_c + succoa_c + thdp_c --> coa_c + sl2a6o_c\n",
      "Enough data for Thiamine transport via ABC system (periplasm) (THMabcpp): atp_c + h2o_c + thm_p --> adp_c + h_c + pi_c + thm_c\n",
      "Enough data for Thiamine transport via diffusion (extracellular to periplasm) (THMtex): thm_e <=> thm_p\n",
      "Enough data for Threonine aldolase (THRA): thr__L_c --> acald_c + gly_c\n",
      "Enough data for L-threonine deaminase (THRD_L): thr__L_c --> 2obut_c + nh4_c\n",
      "Enough data for Threonine synthase (THRS): h2o_c + phom_c --> pi_c + thr__L_c\n",
      "Enough data for L-threonine transport via ABC system (periplasm) (THRabcpp): atp_c + h2o_c + thr__L_p --> adp_c + h_c + pi_c + thr__L_c\n",
      "Enough data for L-threonine transport via diffusion (extracellular to periplasm) (THRtex): thr__L_e <=> thr__L_p\n",
      "Enough data for Thymine transport via diffusion (extracellular to periplasm) (THYMtex): thym_e <=> thym_p\n",
      "Enough data for Transketolase (TKT1): r5p_c + xu5p__D_c <=> g3p_c + s7p_c\n",
      "Enough data for Transketolase (TKT2): e4p_c + xu5p__D_c <=> f6p_c + g3p_c\n",
      "Enough data for Triose-phosphate isomerase (TPI): dhap_c <=> g3p_c\n",
      "Enough data for Tryptophan synthase (indoleglycerol phosphate) (TRPS1): 3ig3p_c + ser__L_c --> g3p_c + h2o_c + trp__L_c\n",
      "Enough data for Tryptophan synthase (indole) (TRPS2): indole_c + ser__L_c --> h2o_c + trp__L_c\n",
      "Enough data for Tryptophan synthase (indoleglycerol phosphate) (TRPS3): 3ig3p_c --> g3p_c + indole_c\n",
      "Enough data for L-tryptophan transport via diffusion (extracellular to periplasm) (TRPtex): trp__L_e <=> trp__L_p\n",
      "Enough data for Tyramine transport via diffusion (extracellular to periplasm) (TYMtex): tym_e <=> tym_p\n",
      "Enough data for Tyrosine transaminase (TYRTA): akg_c + tyr__L_c <=> 34hpp_c + glu__L_c\n",
      "Enough data for L-tyrosine transport via diffusion (extracellular to periplasm) (TYRtex): tyr__L_e <=> tyr__L_p\n",
      "Enough data for UDP-N-acetyl-D-glucosamine pyrophosphohydrolase (periplasm) (UACGAMPpp): h2o_p + uacgam_p --> acgam1p_p + 2.0 h_p + ump_p\n",
      "Enough data for UDP-N-acetyl-D-glucosamine transport via diffusion (extracellular to periplasm) (UACGAMtex): uacgam_e <=> uacgam_p\n",
      "Enough data for UDP-N-acetylglucosamine 1-carboxyvinyltransferase (UAGCVT): pep_c + uacgam_c --> pi_c + uaccg_c\n",
      "Enough data for UDPglucose 4-epimerase (UDPG4E): udpg_c <=> udpgal_c\n",
      "Enough data for UDPgalactose pyrophosphohydrolase (UDPGALPpp): h2o_p + udpgal_p --> gal1p_p + 2.0 h_p + ump_p\n",
      "Enough data for UDPgalactose transport via diffusion (extracellular to periplasm) (UDPGALtex): udpgal_e <=> udpgal_p\n",
      "Enough data for UDPglucose pyrophosphohydrolase (UDPGPpp): h2o_p + udpg_p --> g1p_p + 2.0 h_p + ump_p\n",
      "Enough data for UDPglucose transport via diffusion (extracellular to periplasm) (UDPGtex): udpg_e <=> udpg_p\n",
      "Enough data for UMP kinase (UMPK): atp_c + ump_c <=> adp_c + udp_c\n",
      "Enough data for UMP transport via diffusion (extracellular to periplasm) (UMPtex): ump_e <=> ump_p\n",
      "Enough data for Uracil phosphoribosyltransferase (UPPRT): prpp_c + ura_c --> ppi_c + ump_c\n",
      "Enough data for Uracil transport via diffusion (extracellular to periplasm) (URAtex): ura_e <=> ura_p\n",
      "Enough data for Uridine hydrolase (URIH): h2o_c + uri_c --> rib__D_c + ura_c\n",
      "Enough data for Uridine transport via diffusion (extracellular to periplasm) (URItex): uri_e <=> uri_p\n",
      "Enough data for Valine transaminase (VALTA): akg_c + val__L_c <=> 3mob_c + glu__L_c\n",
      "Enough data for L-valine transport via ABC system (periplasm) (VALabcpp): atp_c + h2o_c + val__L_p --> adp_c + h_c + pi_c + val__L_c\n",
      "Enough data for L-valine transport via diffusion (extracellular to periplasm) (VALtex): val__L_e <=> val__L_p\n",
      "Enough data for Valine-pyruvate aminotransferase (VPAMTr): 3mob_c + ala__L_c <=> pyr_c + val__L_c\n",
      "Enough data for Xanthine transport via diffusion (extracellular to periplasm) (XANtex): xan_e <=> xan_p\n",
      "Enough data for Xanthine phosphoribosyltransferase (XPPT): prpp_c + xan_c --> ppi_c + xmp_c\n"
     ]
    }
   ],
   "source": [
    "enough_data_bn(model,fluxes,proteins2,metabolites)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {
    "collapsed": false
   },
   "source": [
    "Create dataframe including what is the limitation data factor for each reaction:"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 104,
   "metadata": {
    "collapsed": false
   },
   "outputs": [],
   "source": [
    "Name = []\n",
    "Reaction = [] \n",
    "is_flux = [False] * len(model.reactions)\n",
    "is_metab = [False] * len(model.reactions)\n",
    "is_prot = [False] * len(model.reactions)\n",
    "\n",
    "for i in range(len(model.reactions)):\n",
    "    rxn = model.reactions[i]\n",
    "    substrates = rxn.reactants # improvement: consider fluxes\n",
    "    genes = list(rxn.genes) # use gene_reaction_rule to consider and/or\n",
    "\n",
    "    # Check fluxes\n",
    "    flux_bool = False\n",
    "    if any(rxn.id in s for s in list(fluxes.index.values)):\n",
    "        flux_bool = True\n",
    "\n",
    "    # Check proteins\n",
    "    prot_bool = False\n",
    "    for j in range(len(genes)):\n",
    "        prot = mapping[mapping[2]==genes[j].id][0].reset_index()\n",
    "        if prot.empty==0:\n",
    "            gene = list(mapping[(mapping[0]==prot[0][0]) & (mapping[1]=='Gene_Name')][2])\n",
    "            if any(gene[0] in s for s in list(proteins.index.values)):\n",
    "                prot_bool = True\n",
    "                break\n",
    "\n",
    "    # Check substrates\n",
    "    met_bool = False\n",
    "    for j in range(len(substrates)):\n",
    "        if any(substrates[j].id[:-2] in s for s in ['h','h2o'])==0:\n",
    "            if any(substrates[j].id[:-2] in s for s in list(metabolites.index.values))==0:\n",
    "                break\n",
    "    else:\n",
    "        met_bool = True\n",
    "    \n",
    "    Name.append(rxn.name)\n",
    "    Reaction.append(rxn.reaction)\n",
    "    is_flux[i] = flux_bool\n",
    "    is_metab[i] = met_bool\n",
    "    is_prot[i] = prot_bool\n",
    "\n",
    "availability = {'Name':Name, 'Reaction':Reaction, 'is_flux':is_flux, 'is_metab':is_metab, 'is_prot':is_prot}\n",
    "availability = pd.DataFrame(availability)\n",
    "availability.to_excel(data_dir+\"availability.xlsx\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "collapsed": false
   },
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.6.0"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
